src/primitives.rs

```diff
@@ -0,0 +1,133 @@
+use num_derive::FromPrimitive;
+use num_traits::FromPrimitive;
+
+use crate::requests::{BodyApiVersionRequest, BodyMeatdataApiVersionRequest};
+use bytes::Buf;
+use std::fmt::{Error, Formatter};
+
+#[derive(FromPrimitive)]
+pub enum ApiKey {
+    Metadata = 3,
+    ApiVersions = 18,
+}
+
+pub enum ApiVersion {
+    Version0,
+    Version1,
+    Version2,
+}
+
+#[derive(Debug)]
+pub enum DecodeError {
+    BufferUnderflow,
+    BufferOverflow,
+    BadData,
+}
+
+#[derive(Debug, PartialEq)]
+pub struct HeaderRequest {
+    pub api_key: i16,
+    pub api_version: i16,
+    pub correlation_id: i32,
+    pub client_id: String,
+}
+
+impl std::fmt::Display for HeaderRequest {
+    fn fmt(&self, _: &mut Formatter) -> Result<(), Error> {
```

Comment:
You should be writing to the `Formatter` that's passed in, not throwing it away:
```suggestion
    fn fmt(&self, f: &mut Formatter) -> Result<(), Error> {
```